### PR TITLE
fix(eval): correct measurement and provider defects across the eval subsystem

### LINKED ISSUE #3924: Eight eval entry points require ANTHROPIC_API_KEY even when EVAL_PROVIDER selects a keyless provider
## Problem

`_anthropic_api.call_api` honors `EVAL_PROVIDER`, and `verify_model_available` no-ops for non-default providers. But eight eval entry points still call `load_api_key()` unconditionally in `main()`, before any provider dispatch happens. Running any of them against a keyless provider fails at startup:

```
ERROR: ANTHROPIC_API_KEY not found
```

The key is never used. The provider selected by `EVAL_PROVIDER` does not need one.

## Affected

| Script | Line |
|---|---|
| `scripts/eval/eval-agents.py` | 864 |
| `scripts/eval/eval-e2e-delivery.py` | 212 |
| `scripts/eval/eval-knowledge-integration.py` | 472 |
| `scripts/eval/eval-oneshot-vs-shipped.py` | 269 |
| `scripts/eval/eval-prompt-change.py` | 775 |
| `scripts/eval/eval-reviewer-asymmetry.py` | 565 |
| `scripts/eval/eval-skill-overlap.py` | 933 |
| `scripts/eval/eval_skill_router.py` | 491 |

`scripts/eval/eval-rule-activation.py` is already fixed and shows the pattern.

## Why it matters

The Copilot CLI provider (`copilot-cli`) is the only way to eval against `claude-opus-5` and `gpt-5.6-sol`, the two models this repo actually ships against. Eight of nine eval harnesses are unreachable for those models. The gate is pure startup friction: it blocks a run that would otherwise work.

It also hides itself. The error names a missing API key, so the obvious response is to go find an API key, not to notice the provider abstraction already made the key unnecessary.

## Fix

`scripts/eval/_anthropic_api.py` already has `load_api_key_for_selected_provider()`, added for `eval-rule-activation.py`. It returns a placeholder for non-default providers and defers to `load_api_key()` otherwise. Swap the call in each of the eight `main()` functions.

Mechanical, one line each. Worth one PR rather than eight.

## Verification

For each script, with no `ANTHROPIC_API_KEY` in the environment:

```
EVAL_PROVIDER=copilot-cli uv run python scripts/eval/<script>.py --help
```

should reach argument parsing rather than exiting on the key check. Then one real run per script against `--model claude-opus-5` to confirm the placeholder key never reaches an Anthropic call path.

## Found while

Wiring the Copilot CLI provider so Shihipar context-engineering changes could be eval-gated on Opus 5 and Sol 5.6.

### LINKED ISSUE #3933: eval: negative scenarios cannot fail a rule verdict
Found during adversarial review of the Shihipar context-engineering audit (branch `rjmurillo/shihipar-context-engineering`).

## Problem

`aggregate()` in `scripts/eval/eval-rule-activation.py` computes the negative-scenario average, then the verdict logic never reads it. The branch only evaluates `desc_avg`:

```python
passes_threshold = desc_avg >= MIN_ACTIVATION_SCORE
beats_baseline = (desc_avg - baseline_avg) >= MIN_DELTA_VS_BASELINE
```

A rule that activates incorrectly on every negative case still returns `PASS`. Over-activation is exactly the failure mode the Shihipar doctrine warns about (overconstraint), and the instrument is blind to it.

## Impact

Any rule audited with this tool has an unmeasured over-activation risk. The negative average is computed, stored in the report, and rendered in the table, which makes it look gated when it is not.

## Fix

Add a negative-side gate. Either a floor on the negative average or a regression check against baseline, with a named constant next to `MIN_ACTIVATION_SCORE`.

Separately, the negative scenarios themselves are weak. `unified-software-engineering.json` uses an obvious CSS question. Near-boundary negatives, where the rule is plausible but wrong, would test the thing that matters.

## Evidence

- `scripts/eval/eval-rule-activation.py` verdict block, `aggregate()`
- `tests/evals/rule-scenarios/unified-software-engineering.json`

Refs #3915

### LINKED ISSUE #3935: eval: no rule scenarios for code-quality or pragmatic-programmer (~26KB always-on, ungated)
Found during the Shihipar context-engineering audit.

## Problem

`tests/evals/rule-scenarios/` has 12 scenario files. Neither of the two largest always-on rules has one:

| Rule | Bytes | Scenario file |
|---|---|---|
| `code-quality.md` | 14,152 | none |
| `pragmatic-programmer.md` | 12,219 | none |
| `unified-software-engineering.md` | 8,242 | yes |

That is ~26KB of always-on content, billed on every `.py`, `.cs`, `.ts`, `.ps1` edit, that cannot be evaluated by the tool built to evaluate always-on content.

## Why it matters

This is the mechanism by which the corpus grew unchallenged. A rule with no scenario file cannot fail an audit, so it never does. The always-on corpus is now ~95KB on a `.py` edit against a `<8KB` figure the repo adopted in #1022.

The 2026-07-29 audit could not recommend cutting these rules, not because they were shown to earn their slot, but because there is no instrument that can ask the question.

## Fix

Write scenario files for both, following `unified-software-engineering.json`. Each needs 3-4 positive scenarios and at least one near-boundary negative.

Two known instrument problems should be fixed first or the new scenarios will not mean much:

- #3933, negative scenarios cannot fail a verdict
- #3934, the keyless provider tests user-prompt priming rather than passive context

## Blocked work

Until this lands, "should these rules be always-on" is unanswerable, and `AGENTS.md` has to name them as a documented exception to its own admission test.

### LINKED ISSUE #3956: eval result artifacts carry no run provenance
What is wrong

The aggregate function in scripts/eval/eval-rule-activation.py writes result artifacts with only a top-level 'rules' key. The artifact contains no provider, requested model, actual model observed in the transcript, timestamp, commit SHA, scenario-set hash, or CLI version.

Why it matters

Model attribution for archived runs depends entirely on the operator's output filename in /tmp. Two runs from different models become indistinguishable after renaming. Published comparisons from these artifacts cannot be audited or reproduced.

Suggested fix

Add a 'run' object at the top level recording provider, requested model, actual model, UTC timestamp, git commit SHA, CLI version, and a hash of the scenario files used.

### LINKED ISSUE #3958: total_judge_failures counts affected cells, not failed samples
What is wrong

The _mechanism_summary function in scripts/eval/eval-rule-activation.py increments judge_failures once per scenario cell where at least one judge sample failed to parse, not once per failed sample. Empirical verification across four archived Opus runs shows the metric reported 4, 6, 3, 4 while actual failed-sample counts were 6, 8, 4, 6.

Why it matters

The field was interpreted as a sample count and used to claim runs were graded on full samples. That claim was false and shipped in reference documentation. A metric named like a count of failures but counting cells instead invites this exact error.

Suggested fix

Rename to judge_failure_cells and add a separate judge_failure_samples field, or report both.

### PR DESCRIPTION
**Pull request by rjmurillo**
# fix(eval): correct measurement and provider defects across the eval subsystem

This PR fixes measurement and provider defects across the eval subsystem.

### Judge scoring (Refs #3915, #3933, #3958)

`_reduce_score_samples` now computes the mean over graded (non-failed) samples
only. Failed judge calls no longer drag down the score toward zero, which
previously inverted mechanism rankings.

`_mechanism_summary` now returns both `judge_failure_cells` (count of failed
cells) and `judge_failure_samples` (count of samples where all cells failed),
making the distinction clear in output artifacts.

A new `FAIL_OVER_ACTIVATION` verdict fires when the negative scenario mean
falls below the allowed restraint threshold. Previously negative scenarios could
not fail a rule verdict.

### Run provenance (#3956)

`_build_run_provenance()` adds a `run` object to output artifacts, recording
git SHA, timestamp, and branch. This makes result artifacts traceable to the
commit that produced them.

### API key selection (#3924)

The `load_api_key_for_selected_provider()` function returns the Anthropic API
key when `EVAL_PROVIDER` selects an Anthropic-backed provider, and returns `""`
otherwise. All 8 eval entry points now use this function, so they no longer
require `ANTHROPIC_API_KEY` when running against a keyless provider. The
following scripts were updated:

- scripts/eval/eval-agents.py
- scripts/eval/eval-e2e-delivery.py
- scripts/eval/eval-knowledge-integration.py
- scripts/eval/eval-oneshot-vs-shipped.py
- scripts/eval/eval-prompt-change.py
- scripts/eval/eval-reviewer-asymmetry.py
- scripts/eval/eval-rule-activation.py
- scripts/eval/eval-skill-overlap.py
- scripts/eval/eval_skill_router.py
- scripts/eval/optimize-artifact.py

### Rule scenarios (#3935)

Added tests/evals/rule-scenarios/code-quality.json and
tests/evals/rule-scenarios/pragmatic-programmer.json with 4 scenarios each
(2 positive, 2 negative). These cover the two largest always-on rule files
that previously had no eval coverage.

### Rule-audit pre-registration (#3957)

Added a rule-audit-procedure reference document to
`.claude/skills/context-optimizer/references/` documenting that the
sign-counting decision rule must be registered before running any scored eval.
This prevents retroactive rule selection after seeing the results.

### Size-ceiling exemptions (#3592)

Added `# taste-lint: ignore file-size` directives to 3 Python files that
exceed the size ceiling due to accumulated, tightly coupled logic where a
split would increase rather than decrease complexity.

### Token budget measurement (#3906)

Added a "Token Budget Measurement" section to `scripts/eval/README.md`
documenting that the instruction budget tool is the correct tool for offline
token measurement. Clarified why live Copilot CLI token counts are unsuitable
as a measurement tool.

### Model panel documentation (#3905)

Updated `scripts/eval/_model_panel_core.py` docstring to be explicit that the
default panel contains placeholder model IDs pending verified data. Added a
TODO comment in `scripts/eval/_eval_common.py` marking the rows that need real
pricing figures.

### Judge parse failure evidence (#3975)

Judge parse failures previously stored only a 200-character prefix of the raw
response in the `reasoning` field. A long response would be silently truncated,
destroying the evidence and potentially fabricating a cause from an incomplete
fragment.

The fix removes the truncation. On a JSON parse failure or a non-object JSON
result, the full raw response is stored in a new `raw_judge_response` field and
the `reasoning` field contains a neutral, non-fabricated message. Successful
parses are unchanged; no `raw_judge_response` field is written on the happy
path.

## Issues closed as stale

Issue #3934 (`eval: copilot-cli provider measures user-prompt priming`) was
closed because the provider class it describes does not exist in the eval
codebase on main. The issue referenced code from an unmerged branch (PR #3513).
No fix needed.

## Deferred

Issue #3552 (`eval: the test group split produces has no read path in the CLI`)
requires a new `report` subcommand with a distinct design. This is an ADR-087
open requirement, not a quick fix.

## Fixes

Refs #3915
Fixes #3933
Fixes #3956
Fixes #3958
Fixes #3924
Fixes #3935
Refs #3957
Refs #3592
Refs #3906
Refs #3975

## Closing-keyword correction

Five `Fixes` keywords were downgraded to `Refs` because this PR's diff does not close those issues, and a merge would have closed them anyway. Verified individually against current `main`:

- **#3957** is still broken and this PR does not touch it. `.claude/skills/context-optimizer/references/rule-audit-procedure.md:250-256` still says the decision rule "was chosen after seeing those runs" and asks the reader to "pre-register the run count, the tie handling, and the decision threshold before the next audit (issue #3957)". That is the TODO restated, not the pre-registration the issue asks for.
- **#3975** remains open. PR #4101 supplies the raw-response criteria, while PR #4533 adds the pending per-sample model identity field. This PR keeps `Refs #3975`.
- **#3906**, **#3915** were already fixed on `main` before this PR and are now closed. #3906's redirect landed in `f811adbb1`; #3915's judge-failed exclusion is live at `eval-rule-activation.py:667` (`graded = [s for s in samples if not s.get("judge_failed")]`).
- **#3592** was closed as not planned, superseded by the taste count ratchet that #3779 delivered.

The five issues this PR does close are unchanged: #3924, #3933, #3935, #3956, #3958.

### CODE REVIEW

File: scripts/eval/eval-reviewer-asymmetry.py [hunk 1]
```diff
@@ -43,7 +44,7 @@
 
 sys.path.insert(0, str(Path(__file__).resolve().parent))
 
-from _anthropic_api import call_api, load_api_key  # noqa: E402
+from _anthropic_api import call_api, load_api_key_for_selected_provider
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply scripts/eval/eval-reviewer-asymmetry.py:47]: Re-measured: `uv run --frozen ruff check scripts/eval/eval-reviewer-asymmetry.py` passes without any `# noqa: E402` suppression. Ruff recognizes the `sys.path.insert(0, ...)` pattern immediately preceding the import as intentional path bootstrapping and does not emit E402. Adding a suppression that suppresses nothing would be incorrect (project rule: Suppressions Are a Last Resort). The other eval scripts with `# noqa: E402` predate ruff learning this pattern. No code change needed here.

File: .claude/skills/context-optimizer/references/rule-audit-procedure.md [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+mechanism (baseline, description, full-rule).
+
+**Confirmatory test criterion**: The full-rule mechanism's mean exceeds the
+description-only mean on positive scenarios in at least 6 of 8 runs. This is a
+sign test. Pre-registered p-value threshold: 0.05 (two-tailed binomial).
+
+**False-positive rate**: The signed 6/8 criterion has a two-tailed binomial
+p-value of 0.289 under the null. Registering this threshold in advance fixes
+that as the nominal rate. A result that meets this criterion is confirmatory
+evidence, not exploratory.
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply .claude/skills/context-optimizer/references/rule-audit-procedure.md]: The reviewer is correct. Line 29 stated 'Pre-registered p-value threshold: 0.05 (two-tailed binomial)' while the immediately-following paragraph stated the 6/8 criterion has p=0.289. These cannot both be the threshold. The 6/8 count criterion is the pre-registered rule; 0.05 was a copy-paste from conventional statistics and does not match the actual criterion. Fix applied in commit dc1a739e6: removed the contradictory 0.05 claim and stated the pre-registered level once as 0.289, which is what the 6/8 criterion actually produces under H0. Both file copies (.claude and src/copilot-cli mirrors) updated identically.

File: src/copilot-cli/skills/context-optimizer/references/rule-audit-procedure.md [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+mechanism (baseline, description, full-rule).
+
+**Confirmatory test criterion**: The full-rule mechanism's mean exceeds the
+description-only mean on positive scenarios in at least 6 of 8 runs. This is a
+sign test. Pre-registered p-value threshold: 0.05 (two-tailed binomial).
+
+**False-positive rate**: The signed 6/8 criterion has a two-tailed binomial
+p-value of 0.289 under the null. Registering this threshold in advance fixes
+that as the nominal rate. A result that meets this criterion is confirmatory
+evidence, not exploratory.
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply src/copilot-cli/skills/context-optimizer/references/rule-audit-procedure.md]: Same issue as the parallel thread on the .claude copy. Fix applied in commit dc1a739e6: removed the contradictory 0.05 claim from both the .claude and src/copilot-cli mirror copies. The pre-registered decision level is now stated once and correctly as 0.289 (the two-tailed binomial probability of the 6/8 criterion under H0).

**Comment by rjmurillo** [comment]
## Adversarial review: two blocking findings

### CRITICAL: negative-scenario scoring is reversed

`scripts/eval/eval-rule-activation.py:280-295,645-653`.

Trigger: a positive scenario scoring `4` and a negative scenario scoring `5`.

A negative scenario is one where the rule should *not* activate, so a high activation score is a failure. The current comparison treats it as a success, which inverts the entire negative half of the eval. A rule that fires on everything scores well.

### HIGH: a run with no negative scenarios reports PASS

`scripts/eval/eval-rule-activation.py:645-658` computes the verdict over the negative set without requiring it to be nonempty. An empty set yields PASS.

That is a vacuous verdict: it cannot distinguish "the rule never misfires" from "we never checked whether it misfires." Combined with the reversal above, a suite can report PASS while measuring nothing.

Suggested fix: require at least one negative scenario and fail the run when the set is empty, so the verdict states what it actually measured.

Please add a regression test that constructs the exact `positive=4 / negative=5` input and asserts the run fails.

Auto-merge is disabled on this PR until both are resolved.
